azurelinuxagent/common/cgroupconfigurator.py

```diff
@@ -225,29 +237,70 @@ def __impl():
 
-        def get_processes_in_agent_cgroup(self):
+        def check_processes_in_agent_cgroup(self):
             """
-            Returns an array of tuples with the PID and command line of the processes that are currently within the cgroup for the given unit.
-
-            The return value can be None if cgroups are not enabled or if an error occurs during the operation.
+            Verifies that the agent's cgroup includes only the current process, its parent and commands started using shellutil (i.e. the extension handler,
+            the daemon, and the commands started by the extension handler, respectively).
+            Other processes started by the agent (e.g. extensions) and processes not started by the agent (e.g. services installed by extensions) should
+            belong to their own cgroup.
+            The function raises an UnexpectedProcessesInCGroupException if the check fails.
             """
-            def __impl():
-                if self._agent_cpu_cgroup_path is None:
-                    return []
-                return self._cgroups_api.get_processes_in_cgroup(self._agent_cpu_cgroup_path)  # pylint: disable=E1101
-
-            def __on_error(exception):
-                #
-                # Send telemetry for a small sample of errors (if any)
-                #
-                self._get_processes_in_agent_cgroup_error_count = self._get_processes_in_agent_cgroup_error_count + 1
-                if self._get_processes_in_agent_cgroup_error_count <= 5:
-                    message = "Failed to list the processes in the agent's cgroup: {0}", ustr(exception)
-                    if message != self._get_processes_in_agent_cgroup_last_error:
-                        add_event(op=WALAEventOperation.CGroupsDebug, message=message)
-                    self._get_processes_in_agent_cgroup_last_error = message
-
-            return self._invoke_cgroup_operation(__impl, "Failed to list the processes in the agent's cgroup.", on_error=__on_error)
+            if not self.enabled():
+                return
+
+            daemon = os.getppid()
+            extension_handler = os.getpid()
+            agent_commands = set()
+            agent_commands.update(shellutil.get_running_commands())
+            agent_cgroup = CGroupsApi.get_processes_in_cgroup(self._agent_cpu_cgroup_path)
+            # get the running commands again in case a new command was started while we were fetching the processes in the cgroup;
+            agent_commands.update(shellutil.get_running_commands())
+
+            unexpected = []
+            for process in agent_cgroup:
+                if process in (daemon, extension_handler):
+                    continue
+                # check if the process is a command started by the agent or a descendant of one of those commands
+                current = process
+                while current != 0 and current not in agent_commands:
+                    current = self._get_parent(current)
```

Comment:
Why not add `current` but its parent?

Reply:
sorry, not sure i understand the question

Reply:
Didn't realize you are climbing up the process tree until you reach either the agent or the scheduler (parent process of all). Makes sense now!